File: modules/expo-pcm-audio-player/ios/ExpoPcmAudioPlayerModule.swift
import ExpoModulesCore
import AVFoundation

public class ExpoPcmAudioPlayerModule: Module {
  // Audio engine components
  private var audioEngine: AVAudioEngine?
  private var playerNode: AVAudioPlayerNode?
  private var audioFormat: AVAudioFormat?
  
  // Configuration - incoming PCM16 audio specs
  private var inputSampleRate: Double = 16000
  private var inputChannels: Int = 1
  
  // Playback state
  private var isInitialized = false
  private var isPlaying = false
  private var audioQueue: [AVAudioPCMBuffer] = []
  private var queueLock = NSLock()
  
  // Volume
  private var currentVolume: Float = 1.0
  
  public func definition() -> ModuleDefinition {
    Name("ExpoPcmAudioPlayer")
    
    OnCreate {
      // Will be initialized via initialize() call
    }
    
    OnDestroy {
      self.cleanup()
    }
    
    // Initialize audio player
    AsyncFunction("initialize") { (config: [String: Any], promise: Promise) in
      let sampleRateValue = config["sampleRate"] as? Double ?? 16000
      let channelsValue = config["channels"] as? Int ?? 1
      
      self.inputSampleRate = sampleRateValue
      self.inputChannels = channelsValue
      
      do {
        try self.setupAudioEngine()
        self.isInitialized = true
        print("[ExpoPcmAudioPlayer] ✅ Initialized (input: \(sampleRateValue)Hz, \(channelsValue) channel(s), PCM16)")
        promise.resolve(nil)
      } catch {
        print("[ExpoPcmAudioPlayer] ❌ Failed to initialize: \(error)")
        promise.reject("INIT_ERROR", "Failed to initialize: \(error.localizedDescription)")
      }
    }
    
    // Stream audio chunk
    AsyncFunction("streamChunk") { (base64Data: String, promise: Promise) in
      do {
        try self.streamAudioChunk(base64Data: base64Data)
        promise.resolve(nil)
      } catch {
        promise.reject("STREAM_ERROR", "Failed to stream chunk: \(error.localizedDescription)")
      }
    }
    
    // Start playback
    AsyncFunction("start") { (promise: Promise) in
      do {
        try self.startPlayback()
        promise.resolve(nil)
      } catch {
        promise.reject("PLAY_ERROR", "Failed to start playback: \(error.localizedDescription)")
      }
    }
    
    // Pause playback
    AsyncFunction("pause") { (promise: Promise) in
      self.pausePlayback()
      promise.resolve(nil)
    }
    
    // Stop playback
    AsyncFunction("stop") { (promise: Promise) in
      self.stopPlayback()
      promise.resolve(nil)
    }
    
    // Flush buffer
    AsyncFunction("flush") { (promise: Promise) in
      promise.resolve(nil)
    }
    
    // Set volume
    AsyncFunction("setVolume") { (volume: Double, promise: Promise) in
      self.setVolume(volume: Float(volume))
      promise.resolve(nil)
    }
    
    // Get status
    AsyncFunction("getStatus") { (promise: Promise) in
      let status: [String: Any] = [
        "isPlaying": self.isPlaying,
        "buffered": self.audioQueue.count
      ]
      promise.resolve(status)
    }
  }
  
  // MARK: - Audio Engine Setup
  
  private func setupAudioEngine() throws {
    // Clean up existing engine if any
    cleanup()
    
    // Configure audio session FIRST
    let audioSession = AVAudioSession.sharedInstance()
    try audioSession.setCategory(.playback, mode: .default, options: [.mixWithOthers])
    try audioSession.setActive(true)
    print("[ExpoPcmAudioPlayer] ✅ Audio session configured")
    
    // Initialize audio components
    audioEngine = AVAudioEngine()
    playerNode = AVAudioPlayerNode()
    
    guard let engine = audioEngine,
          let node = playerNode else {
      throw NSError(domain: "ExpoPcmAudioPlayer", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to create audio components"])
    }
    
    // Attach player node to engine FIRST
    engine.attach(node)
    
    // Use Float32 format which is what AVAudioEngine prefers
    // We'll convert from Int16 when streaming
    let format = AVAudioFormat(
      standardFormatWithSampleRate: inputSampleRate,
      channels: AVAudioChannelCount(inputChannels)
    )
    
    guard let format = format else {
      throw NSError(domain: "ExpoPcmAudioPlayer", code: -2, userInfo: [NSLocalizedDescriptionKey: "Failed to create audio format"])
    }
    
    audioFormat = format
    print("[ExpoPcmAudioPlayer] 🎼 Using format: \(format)")
    
    // Connect player node to main mixer
    // Using the format we created (Float32, non-interleaved)
    engine.connect(node, to: engine.mainMixerNode, format: format)
    print("[ExpoPcmAudioPlayer] ✅ Connected player node to mixer")
    
    // Start audio engine
    try engine.start()
    print("[ExpoPcmAudioPlayer] ✅ Audio engine started")
    
    // Set initial volume
    node.volume = currentVolume
  }
  
  // MARK: - Audio Streaming
  
  private func streamAudioChunk(base64Data: String) throws {
    guard isInitialized,
          let format = audioFormat,
          let node = playerNode else {
      print("[ExpoPcmAudioPlayer] ❌ streamAudioChunk: Not initialized")
      throw NSError(domain: "ExpoPcmAudioPlayer", code: -1, userInfo: [NSLocalizedDescriptionKey: "Not initialized"])
    }
    
    // Decode base64 to Data
    guard let audioData = Data(base64Encoded: base64Data) else {
      print("[ExpoPcmAudioPlayer] ❌ streamAudioChunk: Invalid base64 data")
      throw NSError(domain: "ExpoPcmAudioPlayer", code: -2, userInfo: [NSLocalizedDescriptionKey: "Invalid base64 data"])
    }
    
    // Calculate frame count from Int16 PCM data
    // Each frame = channelCount * 2 bytes (Int16)
    let bytesPerFrame = inputChannels * 2
    let frameCount = audioData.count / bytesPerFrame
    
    guard frameCount > 0 else {
      print("[ExpoPcmAudioPlayer] ⚠️ No frames in audio data")
      return
    }
    
    print("[ExpoPcmAudioPlayer] 🎵 Received \(audioData.count) bytes → \(frameCount) frames")
    
    // Create Float32 buffer for AVAudioEngine
    guard let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: AVAudioFrameCount(frameCount)) else {
      print("[ExpoPcmAudioPlayer] ❌ Failed to create buffer")
      throw NSError(domain: "ExpoPcmAudioPlayer", code: -3, userInfo: [NSLocalizedDescriptionKey: "Failed to create buffer"])
    }
    
    buffer.frameLength = AVAudioFrameCount(frameCount)
    
    // Convert Int16 PCM to Float32 and copy to buffer
    audioData.withUnsafeBytes { (rawBufferPointer: UnsafeRawBufferPointer) in
      guard let baseAddress = rawBufferPointer.baseAddress else { return }
      let int16Pointer = baseAddress.assumingMemoryBound(to: Int16.self)
      
      guard let floatChannelData = buffer.floatChannelData else { return }
      
      let channelCount = inputChannels
      
      if channelCount == 1 {
        // Mono: Convert Int16 to Float32 directly
        for frame in 0..<frameCount {
          // Convert Int16 (-32768 to 32767) to Float32 (-1.0 to 1.0)
          floatChannelData[0][frame] = Float(int16Pointer[frame]) / 32768.0
        }
      } else {
        // Stereo: De-interleave and convert
        for frame in 0..<frameCount {
          for channel in 0..<channelCount {
            let sampleIndex = frame * channelCount + channel
            floatChannelData[channel][frame] = Float(int16Pointer[sampleIndex]) / 32768.0
          }
        }
      }
    }
    
    // Add to queue and schedule
    queueLock.lock()
    audioQueue.append(buffer)
    let queueSize = audioQueue.count
    queueLock.unlock()
    
    print("[ExpoPcmAudioPlayer] 📥 Queued buffer (queue size: \(queueSize))")
    
    scheduleNextBuffer()
    
    // Auto-start if not playing
    if !isPlaying {
      print("[ExpoPcmAudioPlayer] 🚀 Auto-starting playback")
      try startPlayback()
    }
  }
  
  private func scheduleNextBuffer() {
    guard let node = playerNode else { return }
    
    queueLock.lock()
    defer { queueLock.unlock() }
    
    var scheduledCount = 0
    while !audioQueue.isEmpty {
      let buffer = audioQueue.removeFirst()
      node.scheduleBuffer(buffer) { [weak self] in
        self?.scheduleNextBuffer()
      }
      scheduledCount += 1
    }
    
    if scheduledCount > 0 {
      print("[ExpoPcmAudioPlayer] 🎼 Scheduled \(scheduledCount) buffer(s)")
    }
  }
  
  // MARK: - Playback Control
  
  private func startPlayback() throws {
    guard let node = playerNode else {
      throw NSError(domain: "ExpoPcmAudioPlayer", code: -4, userInfo: [NSLocalizedDescriptionKey: "Player node not initialized"])
    }
    
    if !isPlaying {
      node.play()
      isPlaying = true
      print("[ExpoPcmAudioPlayer] ▶️ Playback started")
    }
  }
  
  private func pausePlayback() {
    guard let node = playerNode else { return }
    
    if isPlaying {
      node.pause()
      isPlaying = false
      print("[ExpoPcmAudioPlayer] ⏸️ Playback paused")
    }
  }
  
  private func stopPlayback() {
    guard let node = playerNode else { return }
    
    node.stop()
    isPlaying = false
    
    queueLock.lock()
    audioQueue.removeAll()
    queueLock.unlock()
    
    print("[ExpoPcmAudioPlayer] 🛑 Playback stopped")
  }
  
  private func setVolume(volume: Float) {
    let clampedVolume = min(max(volume / 100.0, 0.0), 1.0)
    currentVolume = clampedVolume
    
    if let node = playerNode {
      node.volume = clampedVolume
      print("[ExpoPcmAudioPlayer] 🔊 Volume set to \(Int(clampedVolume * 100))%")
    }
  }
  
  // MARK: - Cleanup
  
  private func cleanup() {
    stopPlayback()
    
    if let engine = audioEngine, engine.isRunning {
      engine.stop()
    }
    
    if let node = playerNode, let engine = audioEngine {
      engine.detach(node)
    }
    
    audioEngine = nil
    playerNode = nil
    audioFormat = nil
    isInitialized = false
    print("[ExpoPcmAudioPlayer] 🧹 Cleaned up")
  }
}

File: modules/expo-pcm-audio-player/android/src/main/java/expo/modules/pcmaudioplayer/ExpoPcmAudioPlayerModule.kt
package expo.modules.pcmaudioplayer

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioTrack
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import kotlinx.coroutines.*
import java.util.concurrent.ConcurrentLinkedQueue
import android.util.Base64
import java.nio.ByteBuffer
import java.nio.ByteOrder

class ExpoPcmAudioPlayerModule : Module() {
  // Audio configuration
  private var sampleRate = 16000
  private var channelConfig = AudioFormat.CHANNEL_OUT_MONO
  private val audioFormat = AudioFormat.ENCODING_PCM_16BIT
  
  // AudioTrack instance
  private var audioTrack: AudioTrack? = null
  
  // Queue for audio chunks
  private val audioQueue = ConcurrentLinkedQueue<ByteArray>()
  
  // Playback state
  private var isPlaying = false
  private var isPaused = false
  private var isInitialized = false
  private var playbackJob: Job? = null
  
  // Volume (0.0 to 1.0)
  private var currentVolume = 1.0f
  
  // Scope for coroutines
  private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
  
  override fun definition() = ModuleDefinition {
    Name("ExpoPcmAudioPlayer")
    
    OnCreate {
      // Will be initialized via initialize() call
    }
    
    OnDestroy {
      cleanup()
    }
    
    // Initialize audio player with configuration
    AsyncFunction("initialize") { config: Map<String, Any>, promise: Promise ->
      try {
        val sampleRateValue = (config["sampleRate"] as? Double)?.toInt() ?: 16000
        val channelsValue = (config["channels"] as? Double)?.toInt() ?: 1
        val bitDepthValue = (config["bitDepth"] as? Double)?.toInt() ?: 16
        
        initializeAudioTrack(sampleRateValue, channelsValue, bitDepthValue)
        promise.resolve(null)
      } catch (e: Exception) {
        promise.reject("INIT_ERROR", "Failed to initialize: ${e.message}", e)
      }
    }
    
    // Stream audio chunk
    AsyncFunction("streamChunk") { base64Data: String, promise: Promise ->
      streamAudioChunk(base64Data, promise)
    }
    
    // Start/resume playback
    AsyncFunction("start") { promise: Promise ->
      startPlayback(promise)
    }
    
    // Pause playback
    AsyncFunction("pause") { promise: Promise ->
      try {
        pausePlayback()
        promise.resolve(null)
      } catch (e: Exception) {
        promise.reject("PAUSE_ERROR", "Failed to pause: ${e.message}", e)
      }
    }
    
    // Stop playback
    AsyncFunction("stop") { promise: Promise ->
      try {
        stopPlayback()
        promise.resolve(null)
      } catch (e: Exception) {
        promise.reject("STOP_ERROR", "Failed to stop: ${e.message}", e)
      }
    }
    
    // Flush buffer (no-op, kept for API compatibility)
    AsyncFunction("flush") { promise: Promise ->
      promise.resolve(null)
    }
    
    // Set volume
    AsyncFunction("setVolume") { volume: Double, promise: Promise ->
      try {
        setVolume(volume.toFloat())
        promise.resolve(null)
      } catch (e: Exception) {
        promise.reject("VOLUME_ERROR", "Failed to set volume: ${e.message}", e)
      }
    }
    
    // Get status
    AsyncFunction("getStatus") { promise: Promise ->
      try {
        val status = mapOf(
          "isPlaying" to isPlaying,
          "buffered" to audioQueue.size
        )
        promise.resolve(status)
      } catch (e: Exception) {
        promise.reject("STATUS_ERROR", "Failed to get status: ${e.message}", e)
      }
    }
  }
  
  // MARK: - Audio Track Initialization
  
  private fun initializeAudioTrack(sampleRate: Int, channels: Int, bitDepth: Int) {
    try {
      this.sampleRate = sampleRate
      this.channelConfig = if (channels == 2) AudioFormat.CHANNEL_OUT_STEREO else AudioFormat.CHANNEL_OUT_MONO
      
      val minBufferSize = AudioTrack.getMinBufferSize(
        sampleRate,
        channelConfig,
        audioFormat
      )
      
      audioTrack = AudioTrack(
        AudioAttributes.Builder()
          .setUsage(AudioAttributes.USAGE_MEDIA)
          .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
          .build(),
        AudioFormat.Builder()
          .setSampleRate(sampleRate)
          .setChannelMask(channelConfig)
          .setEncoding(audioFormat)
          .build(),
        minBufferSize * 2,
        AudioTrack.MODE_STREAM,
        android.media.AudioManager.AUDIO_SESSION_ID_GENERATE
      )
      
      audioTrack?.setVolume(currentVolume)
      isInitialized = true
      
      println("[ExpoPcmAudioPlayer] ✅ AudioTrack initialized (sampleRate: $sampleRate, channels: $channels, bitDepth: $bitDepth)")
    } catch (e: Exception) {
      println("[ExpoPcmAudioPlayer] Failed to initialize: ${e.message}")
      throw e
    }
  }
  
  // MARK: - Audio Streaming
  
  private fun streamAudioChunk(base64Data: String, promise: Promise) {
    if (!isInitialized) {
      promise.reject("NOT_INITIALIZED", "Audio player not initialized", null)
      return
    }
    
    try {
      // Decode base64 to byte array
      val audioData = Base64.decode(base64Data, Base64.DEFAULT)
      
      // Add to queue
      audioQueue.offer(audioData)
      
      // Start playback if not already playing and not paused
      if (!isPlaying && !isPaused) {
        startPlaybackInternal()
      }
      
      promise.resolve(null)
    } catch (e: Exception) {
      println("[ExpoPcmAudioPlayer] Error streaming chunk: ${e.message}")
      promise.reject("STREAM_ERROR", "Failed to stream audio chunk: ${e.message}", e)
    }
  }
  
  // MARK: - Playback Control
  
  private fun startPlayback(promise: Promise) {
    try {
      startPlaybackInternal()
      promise.resolve(null)
    } catch (e: Exception) {
      promise.reject("PLAY_ERROR", "Failed to start playback: ${e.message}", e)
    }
  }
  
  private fun startPlaybackInternal() {
    if (isPlaying) return
    
    val track = audioTrack
    if (track == null) {
      println("[ExpoPcmAudioPlayer] AudioTrack not initialized")
      return
    }
    
    isPlaying = true
    isPaused = false
    track.play()
    println("[ExpoPcmAudioPlayer] ▶️ Playback started")
    
    // Start playback loop
    playbackJob = scope.launch {
      while (isPlaying && !isPaused) {
        val data = audioQueue.poll()
        if (data != null) {
          try {
            val track = audioTrack
            if (track == null) {
              println("[ExpoPcmAudioPlayer] AudioTrack not initialized, stopping playback")
              break
            }
            
            val written = track.write(data, 0, data.size)
            if (written < 0) {
              println("[ExpoPcmAudioPlayer] Error writing to AudioTrack: $written")
            }
          } catch (e: Exception) {
            println("[ExpoPcmAudioPlayer] Error in playback loop: ${e.message}")
            break
          }
        } else {
          // No data available, sleep briefly
          delay(10)
        }
      }
    }
  }
  
  private fun pausePlayback() {
    if (!isPlaying) return
    
    isPaused = true
    audioTrack?.pause()
    playbackJob?.cancel()
    playbackJob = null
    
    println("[ExpoPcmAudioPlayer] ⏸️ Playback paused")
  }
  
  private fun stopPlayback() {
    isPlaying = false
    isPaused = false
    
    audioTrack?.pause()
    audioTrack?.flush()
    audioQueue.clear()
    playbackJob?.cancel()
    playbackJob = null
    
    println("[ExpoPcmAudioPlayer] 🛑 Playback stopped")
  }
  
  private fun setVolume(volume: Float) {
    val clampedVolume = volume.coerceIn(0f, 100f) / 100f
    currentVolume = clampedVolume
    
    if (isInitialized) {
      audioTrack?.setVolume(clampedVolume)
      println("[ExpoPcmAudioPlayer] 🔊 Volume set to ${(clampedVolume * 100).toInt()}%")
    }
  }
  
  // MARK: - Cleanup
  
  private fun cleanup() {
    try {
      stopPlaybackInternal()
      audioTrack?.release()
      audioTrack = null
      isInitialized = false
      scope.cancel()
    } catch (e: Exception) {
      println("[ExpoPcmAudioPlayer] Error during cleanup: ${e.message}")
    }
  }
  
  private fun stopPlaybackInternal() {
    isPlaying = false
    isPaused = false
    
    try {
      audioTrack?.pause()
      audioTrack?.flush()
    } catch (e: Exception) {
      println("[ExpoPcmAudioPlayer] Error stopping AudioTrack: ${e.message}")
    }
    
    audioQueue.clear()
    playbackJob?.cancel()
    playbackJob = null
  }
}
